File: src/pyramid.cpp
#include "conversion.h"

using namespace cv;
using namespace Rcpp;

/* Upsampling an image */
Rcpp::NumericMatrix pyramidUp(Rcpp::NumericMatrix imgMat,
                              Rcpp::NumericVector size_, int borderType)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 5);
    cv::Mat outImg;
    std::vector<int> size = as< std::vector<int> >(size_);
    cv::pyrUp(M, outImg, Size(size.at(0), size.at(1)), borderType);
    return (convertMat_CV2RCPP(outImg));
}


/* Downsampling an image */
Rcpp::NumericMatrix pyramidDown(Rcpp::NumericMatrix imgMat,
                                Rcpp::NumericVector size_, int borderType)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 5);
    cv::Mat outImg;
    std::vector<int> size = as< std::vector<int> >(size_);
    cv::pyrDown(M, outImg, Size(size.at(0), size.at(1)), borderType);
    return (convertMat_CV2RCPP(outImg));
}


Rcpp::NumericMatrix pyrMeanShtFlt(Rcpp::NumericMatrix imgMat, double sp, double sr, int maxlevel)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 0);
    cv::Mat outImg;
    cv::pyrMeanShiftFiltering(M, outImg, sp, sr, maxlevel,
                              TermCriteria(TermCriteria::MAX_ITER+TermCriteria::EPS,5,1));
    return (convertMat_CV2RCPP(outImg));
}


RCPP_MODULE(pyramid)
{
    function("pyramidUp", &pyramidUp, "Pyramid upsample");
    function("pyramidDown", &pyramidDown, "Pyramid downsample");
    function("pyrMeanShtFlt", &pyrMeanShtFlt, "Pyramid mean shift filtering");
}

File: R/pyramid.R
##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Pyramid upsample
##' @param img rip object with 1 or 3 channels
##' @param size size of the output image
##' @param borderType border method. Border type 0, i.e. border constant is not supported.
##' @return Downsampled image
##' @author kaustav nandy
pyrUp <- function(img, size, borderType = 4)
{
    mpyramid <- Module("pyramid", PACKAGE = "RIPOpenCV")
    mpyramid$pyramidUp(img, size, borderType)
}

pyrDown <- function(img, size, borderType = 4)
{
    mpyramid <- Module("pyramid", PACKAGE = "RIPOpenCV")
    mpyramid$pyramidDown(img, size, borderType)
}

pyrMeanShiftFiltering <- function(img, sp, sr, maxlevel)
{
    cvdim <- attr(img, 'cvdim')
    if (cvdim[3] != 3)
        error("Only 3-channel images are supported")
    mpyramid <- Module("pyramid", PACKAGE = "RIPOpenCV")
    mpyramid$pyrMeanShtFlt(img, sp, sr, maxlevel)
}
